src/Inventory1.java

```diff
@@ -0,0 +1,198 @@
+
+    /**
+     * Creates initial representation.
+     *
+     * @param size
+     *            - the number of slots this will have
+     */
+    private void createNewRep(int size) {
+        this.slots = new Item[size];
+        this.reqs = new HashSet<String>();
+
+        for (int i = 0; i < size; i++) {
+            this.slots[i] = new BasicItem();
+        }
+    }
+
+    /**
+     * Creates a new Inventory with a single slot.
+     */
+    public Inventory1() {
+        this.createNewRep(1);
+    }
+
+    /**
+     * Create a new Inventory with one or more slots.
+     *
+     * @param size
+     *            the number of slots this will have
+     *
+     * @requires size > 0
+     */
+    public Inventory1(int size) {
+        assert size > 0 : "Violation of size > 0";
+
+        this.createNewRep(size);
+    }
+
+    //CHECKSTYLE: ALLOW THIS METHOD TO BE OVERRIDDEN
+    @Override
+    public void addItem(int slot, Item item) {
+        assert 0 <= slot : "Violation of 0 <= slot";
+        assert slot < this.slots.length : "Violation of slot < |this|";
+        assert item != null : "Violation of item is not null";
+        assert this.slots[slot].isEmpty() || this.slots[slot].equals(
+                item) : "Violation of slot is empty or has an Item of the same name.";
+        assert this.isAllowed(item) : "Violation of isAllowed(item)";
+
+        Item dest = this.slots[slot];
+
+        if (dest.equals(item)) {
+            dest.putTag(Item.COUNT,
+                    dest.tagValue(Item.COUNT) + item.tagValue(Item.COUNT));
+        } else {
+            this.slots[slot] = item;
+        }
+    }
+
+    //CHECKSTYLE: ALLOW THIS METHOD TO BE OVERRIDDEN
+    @Override
+    public Item removeItem(int slot) {
+        assert 0 <= slot : "Violation of 0 <= slot";
+        assert slot < this.slots.length : "Violation of slot < |this|";
+
+        Item removed = this.slots[slot];
+        this.slots[slot] = new BasicItem();
+
+        return removed;
+    }
+
+    //CHECKSTYLE: ALLOW THIS METHOD TO BE OVERRIDDEN
+    @Override
+    public ArrayList<Item> restrict(String tag) {
+        if (!this.reqs.contains(tag)) {
+            this.reqs.add(tag);
+        }
+
+        ArrayList<Item> removed = new ArrayList<>();
+
+        for (int i = 0; i < this.slots.length; i++) {
+
+            if (!this.slots[i].hasTag(tag)) {
+
+                removed.add(this.slots[i]);
+                this.slots[i] = new BasicItem();
+            }
+        }
+
+        return removed;
+    }
+
+    //CHECKSTYLE: ALLOW THIS METHOD TO BE OVERRIDDEN
+    @Override
+    public void freeRestrictions() {
+        this.reqs = new HashSet<>();
+    }
+
+    //CHECKSTYLE: ALLOW THIS METHOD TO BE OVERRIDDEN
+    @Override
+    public boolean isAllowed(Item item) {
+        assert item != null : "Violation of item is not null";
+
+        boolean allow = true;
+
+        for (String t : this.reqs) {
+            if (!item.hasTag(t)) {
+                allow = false;
```

Comment:
This would be a good candidate for multiple returns. You could return false here the moment you find a missing tag. Unless, the if afterwards saves the day.